### ISSUE TITLE
Tasks to Complete for .NET Core Support

### ISSUE THREAD

**Issue by NightOwl888** [ikvm-revived/ikvm#40]
This is the **official IKVM repository** where efforts are being consolidated to add support for .NET Core/.NET 5/.NET 6, however **we don't currently have a package on NuGet.org**, which is a blocker for many people.

The latest release of IKVM is [the package released by Windward Studios](https://www.windwardstudios.com/blog/ikvm-is-alive-well) and is [available on NuGet.org](https://www.nuget.org/packages/IKVM.WINDWARD/), however efforts to continue that work to port to .NET Core have been moved here.

- [x] We don't own the official [IKVM package](https://www.nuget.org/packages/IKVM/) on NuGet.org (although we do have [permission to use the IKVM name](https://github.com/ikvm-revived/ikvm/issues/13#issuecomment-590334796)). @jfrijters - could you provide us with access to upload a new [IKVM package](https://www.nuget.org/packages/IKVM/) on NuGet.org?
- [x] There is currently not a stable build that work for general use that we can package and release on NuGet.org.
  - At this point we would be willing to forego support for .NET Standard and move straight to .NET 5/6.
  - Ideally, there would still be a target for .NET Framework 4.5+ in the package so consumers don't have to target 2 different IKVM versions to multi-target.
  - [x] This is blocked by the fact that the library structure keeps changing and to target each version of .NET Core, there needs to be a mechanism that can detect which external libraries are required so they can be referenced in the build.
    - There are some efforts to complete this work in https://github.com/ikvm-revived/ikvm/pull/23, https://github.com/ikvm-revived/ikvm/pull/25, and https://github.com/ikvm-revived/ikvm/pull/38.
- [x] Strong naming is incomplete
  - [x] We need to either get the original strong name key (preferred) or generate a new one for IKVM. See https://github.com/ikvm-revived/ikvm/issues/40#issuecomment-1019301101.
  - [x] The strong name key is not in the public repo.
- [x] NuGet Packaging is incomplete
  - We have already migrated to the new .csproj format, so this can be done by [adding the MSBuild attributes](https://docs.microsoft.com/en-us/nuget/create-packages/creating-a-package-msbuild)
- [x] Set up build automation piplines
   - [x] Set up official Azure DevOps organization for ikvm-revivied - Now set up at https://dev.azure.com/ikvm-revived/
      - Note that I already have several multi-targeted class library builds set up on Azure DevOps that we can borrow configs from. See https://github.com/ikvm-revived/ikvm/issues/40#issuecomment-1019325696.
- [x] Documentation on how to build needs to be updated (see #22).

It would be appreciated if others working on ikvm-revived help keep this list updated with the latest info (wiki style) so those willing to contribute know where to focus their efforts.

## Discord

We have set up a Discord server to collaborate on the remaining work: https://discord.gg/MpzNd5Tk8P

**Comment by jfrijters** [comment]
I didn't create the IKVM package on NuGet.org. That was done by barchen. Could you please ask barchen for access first?

**Comment by NightOwl888** [comment]
@jfrijters - Thanks. I used the contact link on NuGet.org to try to get in touch with him. However, as long as you have administrative permissions there, you should be able to assign additional people as administrators. If you assign me, I can set up an ikvm-revived group so other members of the team can upload packages.

**Comment by ripenko** [comment]
I added you as owner of IKVM.

**Comment by NightOwl888** [comment]
@ripenko - Thank you.

**Comment by NightOwl888** [comment]
Okay, now we are over that hurdle. If we are using the same name and same NuGet.org feed, it also follows to use the same strong name key as the identity of the package author. I haven't attempted to look for the original `.snk` file yet, but does everyone agree that is the correct decision?

**Comment by joniles** [comment]
Looking at the notes on strong names [here](https://docs.microsoft.com/en-us/dotnet/standard/assembly/strong-named), I'm not sure we'd lose anything by starting with a fresh `.snk`.

On a related topic, the notes also offer this advice:

> If you are an open-source developer and you want the identity benefits of a strong-named assembly for better compatibility with .NET Framework, consider checking in the private key associated with an assembly to your source control system.

which would definitely make builds simpler.

**Comment by AliveDevil** [comment]
We (as in [our organization](https://github.com/iterate-ch/)) would be open to provide the [JDK8u312b07](https://github.com/iterate-ch/ikvm/commit/6b62850deb12350ce08aa5a1361bde127c972f77) update, once IKVM builds for .NET Framework and .NET 5/6.

In the current state I cannot get either the `master` nor the `ikvm-build`-branch to build anything successfully.

**Comment by NightOwl888** [comment]
> Looking at the notes on strong names here, I'm not sure we'd lose anything by starting with a fresh .snk.

The main issue with *not* using the same key is for existing libraries that specify IKVM as a dependency. For example if an existing library has a package reference to IKVM 8.1.5717:

```xml
<PackageReference Include="IKVM" Version="8.1.5717" />
```

What this effectively means is:

```
IKVM (>= 8.1.5717)
```

If a different `.snk` is used for IKVM 9.0 and the consumer of the library that referenced 8.1.5717 decides to upgrade to IKVM 9.0, they are in for a nasty surprise because even if the API is exactly the same, the different `.snk` file means they are not binary compatible. The upgrade fails, and the only resolutions are:

1. Downgrade to 8.1.5717
2. Petition the owner of the library that referenced 8.1.5717 to upgrade to 9.0

This of course puts the owner of the library at odds because if they upgrade to 9.0 it may break some applications and if they don't upgrade other applications won't be able to upgrade IKVM. And there is no realistic solution that works for everyone.

This is moot if we actually *will* have breaking API changes, but we should avoid that if it is possible.

> On a related topic, the notes also offer this advice:

> > If you are an open-source developer and you want the identity benefits of a strong-named assembly for better compatibility with .NET Framework, consider checking in the private key associated with an assembly to your source control system.

> which would definitely make builds simpler.

Agreed.

**Comment by NightOwl888** [comment]
> We (as in our organization) would be open to provide the JDK8u312b07 update, once IKVM builds for .NET Framework and .NET 5/6.

> In the current state I cannot get either the master nor the ikvm-build-branch to build anything successfully.

Thanks. Are you saying the JDK version affects your ability to build or if you are making 2 unrelated (but helpful) comments?

**Comment by NightOwl888** [comment]
@jfrijters, @ripenko

Since ikvm.net has gone offline, we are now missing binaries and documentation that could be valuable resources going forward. Is there any chance that one of you has a backup of the website files that we can put into a separate GitHub repo in the ikvm-revived organization for reference purposes?

**Comment by jfrijters** [comment]
My ISP decided they wanted back the IPv4 subnet they were selling me and since (ICYMI) I have metastatic prostate cancer the site was going to go down eventually anyway and because I have different priorities nowadays I didn't bother looking for an alternative.

I still have the binaries. Is there anything specific you are looking for?

**Comment by ams-tschoening** [comment]
> [...]But if there is a backup of the binaries that you were hosting, it would definitely be helpful for support purposes.

I have collected everything I found somewhat useful when dealing with IKVM. Feel free to have a look and fork or I might transfer repos into ikvm-revived and fork myself. Whatever makes more sense...

https://github.com/ams-ts-ikvm/ikvm-bin
https://github.com/ams-ts-ikvm

**Comment by ApaulMyLittleAirport** [comment]
How long will it take to migrate to.Netcore?

**Comment by NightOwl888** [comment]
@ApaulMyLittleAirport 

The main issues are:

- Nobody is currently working on it
- Those who wish to contribute are confused about how to contribute (which branch to fork to help out)

I opened this issue because there are several people willing to help, but they are having trouble navigating the confusing branch and repository history of this project since it stopped being maintained by @jfrijters.

While I can say for sure that the [`master` branch](https://github.com/ikvm-revived/ikvm/tree/master) is where we are doing the work, we have some open PRs (#23 and #25) but it isn't clear whether these are ready to be merged or whether there is more to be done on them before they can be merged. It would certainly help if someone would review these PRs. Unfortunately, I haven't looked deep enough into the code to make this determination myself and don't have much bandwidth to contribute here.

**Comment by asbjornu** [comment]
@NightOwl888, thanks for the clarification. If people agree that this is now the official repository for future work on IKVM, perhaps a PR that updates the `README.md` file with this information and removal of all irrelevant and outdated information would be a good idea?

As for pull requests with unclear status; adding some status checks with GitHub Actions would greatly improve that situation. The code should at least be compiled in an action to ensure that a PR doesn't break anything. Unless we are currently in a state where this thing doesn't even build?

I'm sure there are people (including me) that can help with tasks such as this, but it's still difficult to understand what needs to be done, in which order and to figure out the general direction of this project.

**Comment by joniles** [comment]
I've built IKVM locally to get a .Net Core version of my own project and it works beautifully. (I used a strong name key I generated and a couple of manual updates to the build files).

As per the list of tasks at the top of this conversation, we're relatively close to being able to formally release a new built. I'm happy to put time into getting these tasks done.

We should probably focus our efforts on one thing at at time. The first task I'd like to try and tick off is the `snk` file. As per [Microsoft's recommendation](https://docs.microsoft.com/en-us/dotnet/standard/assembly/strong-named), I'd like to get an `snk` file checked in to this repo.

To that end, we need to make a decision about compatibility. @jfrijters you've been very generous with your time in responding here, and I don't want to impose: if you do have the original `snk` file and would be comfortable posting it here that should make for a smoother upgrade process for IKVM users. If that's not possible, I will generate a new `snk` file and create a PR to include it here. In that case we'll have to accept that a minority. of users _may_ face compatibility issues as projects update to use newer versions of IKVM, and move on.

I'm very excited about getting this moving forward!

**Comment by jfrijters** [comment]
[ikvm-key.zip](https://github.com/ikvm-revived/ikvm/files/7918710/ikvm-key.zip)
It's been long enough since Microsoft deprecated using strong names for security that I think it is fine to share my key. So here it is.

**Comment by NightOwl888** [comment]
> I've built IKVM locally to get a .Net Core version of my own project and it works beautifully. (I used a strong name key I generated and a couple of manual updates to the build files).
> 
> As per the list of tasks at the top of this conversation, we're relatively close to being able to formally release a new built. I'm happy to put time into getting these tasks done.
> 
> We should probably focus our efforts on one thing at at time. The first task I'd like to try and tick off is the `snk` file. As per [Microsoft's recommendation](https://docs.microsoft.com/en-us/dotnet/standard/assembly/strong-named), I'd like to get an `snk` file checked in to this repo.
> 
> To that end, we need to make a decision about compatibility. @jfrijters you've been very generous with your time in responding here, and I don't want to impose: if you do have the original `snk` file and would be comfortable posting it here that should make for a smoother upgrade process for IKVM users. If that's not possible, I will generate a new `snk` file and create a PR to include it here. In that case we'll have to accept that a minority. of users _may_ face compatibility issues as projects update to use newer versions of IKVM, and move on.
> 
> I'm very excited about getting this moving forward!

@joniles - That's great. Thanks for stepping up.

First of all, when you say you "built it for .NET Core", which branch(es) did you build? And which version of .NET Core?

I have developed a common solution setup for class libraries that I use on 6 different projects. I integrated some of the features  here, but there are still other features that can be migrated here. Here is a list (off the top of my head).

- [Source Link](https://github.com/dotnet/sourcelink) support
- [Nerdbank.GitVersioning](https://github.com/dotnet/Nerdbank.GitVersioning) to version dev builds based on the commit hash (patched in [version.ps1](https://github.com/Spatial4n/Spatial4n/blob/main/.build/version.ps1) to support 4 segment version numbers)
- InternalsVisibleTo generation via MSBuild (switching to public key hash if it exists in [`<PublicKey>`](https://github.com/ikvm-revived/ikvm/blob/a9989ea1430a5dfd42154b17a49505a5904471b8/Directory.Build.props#L15)). Storing the public key in the MSBuild file eliminates the dependency on `sn.exe`, so the build will work on any OS.
- [dependencies.props](https://github.com/ikvm-revived/ikvm/blob/a9989ea1430a5dfd42154b17a49505a5904471b8/build/dependencies.props) contains all NuGet package versions for the entire solution
- [nuget.props](https://github.com/ikvm-revived/ikvm/blob/a9989ea1430a5dfd42154b17a49505a5904471b8/build/nuget.props) contains most common NuGet package settings
- [Reference Assemblies](https://www.nuget.org/packages/Microsoft.NETFramework.ReferenceAssemblies/) added globally, so we can build without the .NET Framework Developer Pack
- Azure DevOps templates for building, publishing, testing on multiple operating systems in parallel, and reporting any errors under a heading that includes target framework, platform, and OS
- Bootstrap `build.bat` and `build` (for BASH)
- PowerShell `build.ps1` for parsing parameters, choosing defaults, and calling `runbuild.ps1`, which uses Psake build automation to do a local build and test
- How to build documentation (*this will need to be customized for this project)
- How to release documentation

I agree we should tick things off one item at a time. You are welcome to migrate these build features from the following projects:

- [RandomizedTesting](https://github.com/NightOwl888/RandomizedTesting/)
- [ICU4N](https://github.com/NightOwl888/ICU4N/)
- [J2N](https://github.com/NightOwl888/J2N)
- [Morfologik.Stemming](https://github.com/NightOwl888/Morfologik.Stemming)
- [Spatial4n](https://github.com/Spatial4n/Spatial4n)
- [Lucene.Net](https://github.com/apache/lucenenet)

Let me know if you need assistance migrating any of these features or if you feel something is out of your depth.

**Comment by NightOwl888** [comment]
> @NightOwl888, thanks for the clarification. If people agree that this is now the official repository for future work on IKVM, perhaps a PR that updates the `README.md` file with this information and removal of all irrelevant and outdated information would be a good idea?
> 
> As for pull requests with unclear status; adding some status checks with GitHub Actions would greatly improve that situation. The code should at least be compiled in an action to ensure that a PR doesn't break anything. Unless we are currently in a state where this thing doesn't even build?
> 
> I'm sure there are people (including me) that can help with tasks such as this, but it's still difficult to understand what needs to be done, in which order and to figure out the general direction of this project.

@asbjornu 

Well, this is just as much of a discovery for me as it is for you. Several people have provided feedback that they were able to create a local build, but they unfortunately have never elaborated which branch they built (or branches they merged) or provided any kind of instructions, so I am in the dark just as much as you are. Unfortunately, I can't provide info that I don't have, and I am hesitant to remove information that might still have relevance from the README, as it may be required info to come up with the "new" way of doing things even if it doesn't currently work. Of course, if someone were willing to do the research, update documentation, or submit other PRs to bring this project forward such as investigating more about the specifics of what needs to be done, they are welcome to. But please don't wait for me to do the research if you expect a release soon, as I have my hands full with other projects and consider this dependency to be very low in priority at this point.

As for setting up status checks, we need a working build first before we can put in the automation. I have recently setup an Azure DevOps organization at https://dev.azure.com/ikvm-revived/ and completed the integration and approval steps for parallel jobs, so we only need to bring the templates over from [our other projects](https://github.com/ikvm-revived/ikvm/issues/40#issuecomment-1019325696) once we have a build. And of course, @ripenko was kind enough to provide permission to the NuGet.org package so it can be uploaded and and @jfrijters was kind enough to provide the snk file for the build.

**Comment by NightOwl888** [comment]
@joniles - Thanks.

So, if I understand correctly, #25 and #38 can be merged, correct?

Although, I took a peek at #25 and noticed that the `<InternalsVisibleTo>` MSBuild settings were changed so they may not automatically output the public key when the snk is added. Did you have to account for this when you did a build?

The only one of [the open tasks](https://github.com/ikvm-revived/ikvm/pull/25#issue-706325787) I am unsure about is "building missing parts for the core", but I suppose putting a prerelease on NuGet will start people reporting what isn't working, which can only help move us along.

**Comment by joniles** [comment]
@NightOwl888: yes https://github.com/ikvm-revived/ikvm/pull/38 is definitely good to go. With regard to https://github.com/ikvm-revived/ikvm/pull/25, do we want to keep developing against/merging into this branch until we're ready to release, or are we happy to start working against `master` now? Using `master` feels like it would reduce confusion about the direction this repo is heading in!

Regarding `<InternalsVisibleTo>` and use of the `snk` generally, I made a hacky change to the build locally (which you can see [here](https://github.com/ikvm-revived/ikvm/compare/ikvm-build...joniles:ikvm-build-jdbc-signed?expand=1)) to get the strong name stuff working. I was going to open a PR for it, but I then realised that some of the work already seemed to be in place in some of the build files, so I dropped it. I was going to revisit this in the PR which adds the `snk` file.

**Comment by NightOwl888** [comment]
> @NightOwl888: yes #38 is definitely good to go. With regard to #25, do we want to keep developing against/merging into this branch until we're ready to release, or are we happy to start working against `master` now? Using `master` feels like it would reduce confusion about the direction this repo is heading in!

Agreed I think it would be better to keep PRs short lived so we can pool resources on `master`.

So are we good to go right now? Or do you have open work that targets one of these branches?

> Regarding `<InternalsVisibleTo>` and use of the `snk` generally, I made a hacky change to the build locally (which you can see [here](https://github.com/ikvm-revived/ikvm/compare/ikvm-build...joniles:ikvm-build-jdbc-signed?expand=1)) to get the strong name stuff working. I was going to open a PR for it, but I then realised that some of the work already seemed to be in place in some of the build files, so I dropped it. I was going to revisit this in the PR which adds the `snk` file.

Alright, that makes sense. 

The way it was set up here originally, the [root `Directory.Build.targets` file](https://github.com/ikvm-revived/ikvm/blob/a9989ea1430a5dfd42154b17a49505a5904471b8/Directory.Build.targets#L24-L32) has a target to automatically output the attributes during the build. But one of the .NET SDK releases broke this target because the event no longer fires at the right time. It was originally setup like

```
BeforeTargets="CoreCompile"
```

but now that needs to be changed to

```
BeforeTargets="BeforeCompile"
```

Then each project can simply add a list like below, and the attributes will be output automatically.

```
  <ItemGroup>
    <InternalsVisibleTo Include="IKVM.Runtime.JNI" />
    <InternalsVisibleTo Include="IKVM.OpenJDK.Core" />
    <InternalsVisibleTo Include="IKVM.OpenJDK.Util" />
    <InternalsVisibleTo Include="IKVM.OpenJDK.Security" />
    <InternalsVisibleTo Include="IKVM.OpenJDK.Management" />
    <InternalsVisibleTo Include="IKVM.OpenJDK.Media" />
    <InternalsVisibleTo Include="IKVM.OpenJDK.Misc" />
    <InternalsVisibleTo Include="IKVM.OpenJDK.Remoting" />
    <InternalsVisibleTo Include="IKVM.OpenJDK.SwingAWT" />
  </ItemGroup>
```

Does `build.core.cmd` need to own the sn key? Or can we just set the MSBuild properties in the root `Directory.Build.props` file?

### Dotnet Tools

Some of the command line tools either were (or could be) converted to [dotnet tool](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-tool-install)s so they can be installed a NuGet package simply by using the dotnet commands. While we should take things one step at a time, this pertains to setting up the NuGet packaging, so should be taken into account when that is done. There are examples [here](https://www.nuget.org/packages?packagetype=dotnettool) of dotnet tools. Converting a command line tool to a dotnet tool generally isn't very complicated and allows leveraging the same deployment mechanism as the DLLs so the build and deployment is simpler.

https://docs.microsoft.com/en-us/dotnet/core/tools/global-tools-how-to-create

That being said, I am 100% sure that will work on the .NET Core side, but not sure whether there are any blocking issues that will prevent a .NET Core-based command line tool to build the .NET Framework bits, or whether we will need a separate assembly for .NET Framework. Do you think you could find an answer to whether a single set of tools will work for building IKVM assemblies on both .NET Framework and .NET Core or whether separate assemblies are required?

I also just discovered that it is [possible to multi-target dotnet tools](https://twitter.com/kathleendollard/status/1079811275641696256), which I have confirmed several of the projects on the above list are doing. So, we can add targets so the tools will work across major versions of .NET depending on which version is installed on the target machine. Of course, I suspect this doesn't have any .NET Framework support, but if we need a separate tool version for .NET Framework, we might be able to include it in the package, anyway.

**Comment by joniles** [comment]
> So are we good to go right now? Or do you have open work that targets one of these branches?

I think so, I don't have any further changes to make to those branches.

> Does `build.core.cmd need` to own the sn key? Or can we just set the MSBuild properties in the root `Directory.Build.props` file?

I think keeping the `build.*.cmd` scripts as simple as possible is preferable, so if we can add the snk details as properties and reference those that would be great.

> Dotnet Tools

This is an interesting question. For my own use case when I'm packaging my project for distribution,  I only really need the command line tools, which historically I'd have retrieved and unpacked from a zip file, so I'm hoping we retain that option and host simple zipped up releases here on GitHub. Arguably retrieving the IKVM tools as part of a Nuget package doesn't offer me much, particularly if there is an extra step for me to install the tools from the package before I can use them. Stable Nuget packages for me will be the big benefit for my downstream consumers as I won't have to bundle the IKVM assemblies, I can just reference them from the package I build. I appreciate my use case may be a bit odd though!

**Comment by NightOwl888** [comment]
### Dotnet Tools
> 
> This is an interesting question. For my own use case when I'm packaging my project for distribution, I only really need the command line tools, which historically I'd have retrieved and unpacked from a zip file, so I'm hoping we retain that option and host simple zipped up releases here on GitHub. Arguably retrieving the IKVM tools as part of a Nuget package doesn't offer me much, particularly if there is an extra step for me to install the tools from the package before I can use them. Stable Nuget packages for me will be the big benefit for my downstream consumers as I won't have to bundle the IKVM assemblies, I can just reference them from the package I build. I appreciate my use case may be a bit odd though!

Actually, I am not sure whether putting the tool into the IKVM NuGet package is possible (I guess nothing would prevent it if you wanted to do that, though). What I meant was create 1 NuGet package per executable. In general, it is usually best to keep 1 DLL in each NuGet package also, but maybe that wouldn't be the best approach for this project.

Of course, to convert it to a dotnet tool, it would make more sense to have 1 executable that contains several different commands rather than having multiple executables. Then every command could be `ikvm <command> [<subcommand>] [options]`. We used a [template that Microsoft created](https://github.com/natemcmaster/CommandLineUtils) to [build this tool](https://github.com/apache/lucenenet/tree/master/src/dotnet/tools/lucene-cli) with [multiple commands and subcommands](https://lucenenet.apache.org/docs/4.8.0-beta00015/cli/index.html). Having multiple executables is fine, but the downside is the user would have to do multiple installations and we would have to maintain multiple NuGet packages.

The workflow doesn't require you to install the tool every time you use it. You can install it globally on the machine and then use it as many times as you like. I don't know whether this project will drift enough to require the versions to match between executable and DLLs, but it is something that can probably be automated.

A huge advantage of using a dotnet tool is that automation is simpler. Since the `dotnet tool install` command knows how to download the package (including switching TLS version) and unzip it to the install location, there is much less code to write to install it on a build machine than to download it from a GitHub release - just 1 line. And it is possible to write scripts to install the tool locally so it won't crash if the tool is already installed globally and even uninstall the tool to clean up. [See this example](https://github.com/Spatial4n/Spatial4n/blob/main/.build/version.ps1#L76). There is no need to work out which OS you are downloading the tool for and selecting the version is as simple as including the `--version` parameter. And there is no need to check the binaries into your repo to make them part of your build.

As for downloading the binaries:

1. `.nupkg` files are simply zip files with a different extension, so you could technically download the package from NuGet.org and unzip it yourself if you wanted a manual install.
2. The binaries will also be available as artifacts on the Azure DevOps build pipeline, and they are public so anyone can download them.

Keep in mind that executables on .NET Core are no longer xCopy - they need to be [published](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-publish) before they can be deployed. The dotnet pack command automatically makes the tool portable, so no publish is required.

Of course there is nothing wrong with publishing the files and including them for download on the GitHub releases page for those who want to use the old school download approach, it is just extra steps for the release procedure. Unfortunately, while the APIs are there to automate creating a GitHub release, I haven't yet discovered a way to automate "release notes" (as the commit log usually contains commits that need to be summarized better than the commit message), so for now I do that step manually. But we could probably add a build step to automate a GitHub draft release that pushes all of the relevant binaries so it doesn't take extra manual work to upload them to GitHub.

### InternalsVisibleTo

I am in the process of creating a PR to fix this on the ikvm-build branch before it is merged. There was another condition that was added since I contributed it that I didn't mention above which prevents it from firing for packages that don't have `<SignAssembly>` set to `true`, which makes it possible to have unsigned assemblies in the solution.

**Comment by viceice** [comment]
We have a really cool bot, which can manage and suggest updates the installed versions you reference in your projects.

Checkout [renovate](/renovatebot/renovate), i use it for daily work for our dotnet and npm projects. It can do more languages too.

So to install your simply do `dotnet tool restore && dotnet restore` on ci to have all build stuff available.

Renovate will just send you Pr's (like dependabot) to suggest the right updates.

**Comment by NightOwl888** [comment]
@joniles 

I have submitted #41 with some updates, but unfortunately specifying an snk using the file path doesn't seem to function like it does when putting the snk in the keystore. I couldn't find a solution. But I have confirmed all but the [last commit](https://github.com/ikvm-revived/ikvm/pull/41/commits/f4149fc1348dea1b24f3bcbf5e7380319cddd6d7) (enabling the snk) will compile on .NET Core by checking out the one prior.

This post was somewhat helpful: https://ikvm-developers.narkive.com/0VenMex7/ikvmc-cannot-sign-with-strong-name-key

I also noticed that the `build.framework.cmd` is complaining about missing runtime identifiers, but didn't do much investigation on how to resolve it.

I will leave this in your capable hands. Please base your changes on the [ikvm-build-2](https://github.com/ikvm-revived/ikvm/tree/ikvm-build-2) branch and target it when submitting a PR. I will merge everything back to master when we have functioning builds on both .NET Core and .NET Framework.

### Dotnet Tools

After poking around a bit, I can see this is going to take more work than I thought. Even using the simple approach of using the [template](https://github.com/natemcmaster/CommandLineUtils) as a wrapper to convert parameters from the newer format to the original won't quite work because I noticed that some fields that expect file paths don't seem to have a way to escape if they have a space in the path.

I still think it is worth the effort to do this, but I concede that it isn't required for the first prerelease(s) that we publish. It is something that can be added to the build later.

**Comment by AliveDevil** [comment]
There is still so much more to do to support .NET Core, than just getting it build on .NET Core.
The whole dynamic-assembly modules need to be reworked in order to interop correctly with core.

I.e. Log4j crashes completely while setting up the LoggerContext:
```
Unhandled exception. System.TypeInitializationException: The type initializer for 'ch.cyberduck.core.preferences.DefaultPreferences' threw an exception.
 ---> System.TypeInitializationException: The type initializer for 'ch.cyberduck.core.preferences.Preferences' threw an exception.
 ---> System.TypeInitializationException: The type initializer for 'org.apache.logging.log4j.core.config.ConfigurationFactory' threw an exception.
 ---> System.NotSupportedException: The invoked member is not supported in a dynamic assembly.
   at Java_java_lang_Class.forName0(String name, Boolean initialize, ClassLoader loader, Class caller) in ikvm-revived\runtime\IKVM.Runtime\openjdk\java.lang.cs:line 76
   at java.lang.Class.forName0(String name, Boolean initialize, ClassLoader loader, Class caller)
   at java.lang.Class.forName(String className, CallerID )
   at org.apache.logging.log4j.util.LoaderUtil.loadClass(String className)
   at org.apache.logging.log4j.util.LoaderUtil.newInstanceOf(String className)
   at org.apache.logging.log4j.util.LoaderUtil.newCheckedInstanceOf(String className, Class clazz)
   at org.apache.logging.log4j.core.util.Loader.newCheckedInstanceOf(String className, Class clazz)
   at org.apache.logging.log4j.core.lookup.Interpolator..ctor(Map properties)
   at org.apache.logging.log4j.core.lookup.Interpolator..ctor()
   at org.apache.logging.log4j.core.config.ConfigurationFactory..ctor()
   at org.apache.logging.log4j.core.config.ConfigurationFactory.Factory..ctor()
   at org.apache.logging.log4j.core.config.ConfigurationFactory.Factory..ctor(1 x0)
   at org.apache.logging.log4j.core.config.ConfigurationFactory..cctor()
   --- End of inner exception stack trace ---
   at org.apache.logging.log4j.core.config.ConfigurationFactory.getInstance()
   at org.apache.logging.log4j.core.LoggerContext.reconfigure(URI configURI)
   at org.apache.logging.log4j.core.LoggerContext.reconfigure()
   at org.apache.logging.log4j.core.LoggerContext.start()
   at org.apache.logging.log4j.core.impl.Log4jContextFactory.getContext(String fqcn, ClassLoader loader, Object externalContext, Boolean currentContext)
   at org.apache.logging.log4j.core.impl.Log4jContextFactory.<bridge>getContext(String fqcn, ClassLoader loader, Object externalContext, Boolean currentContext)
   at org.apache.logging.log4j.core.impl.Log4jContextFactory.__<overridestub>org.apache.logging.log4j.spi.LoggerContextFactory::getContext(String , ClassLoader , Object , Boolean )
   at org.apache.logging.log4j.LogManager.getContext(ClassLoader loader, Boolean currentContext)
   at org.apache.logging.log4j.LogManager.getLogger(Class clazz)
   at ch.cyberduck.core.preferences.Preferences..cctor()
   --- End of inner exception stack trace ---
   at ch.cyberduck.core.preferences.Preferences.__<clinit>()
   at ch.cyberduck.core.preferences.DefaultPreferences..cctor()
   --- End of inner exception stack trace ---
   at ch.cyberduck.core.preferences.DefaultPreferences..ctor()
   at Ch.Cyberduck.Core.Preferences.AppConfigPreferences..ctor(Locales locales)
   at Ch.Cyberduck.Cli.WindowsTerminalPreferences..ctor() in cyberduck\cli\src\main\csharp\ch\cyberduck\cli\WindowsTerminalPreferences.cs:line 36
   at Ch.Cyberduck.Cli.WindowsTerminal.Main(String[] args) in cyberduck\cli\src\main\csharp\ch\cyberduck\cli\WindowsTerminal.cs:line 38
```

Offending java source: https://github.com/apache/logging-log4j2/blob/rel/2.17.1/log4j-core/src/main/java/org/apache/logging/log4j/core/lookup/Interpolator.java#L117

**Comment by ams-tschoening** [comment]
You should document that as individual issue and link here only.

**Comment by hfickes** [comment]
Is there a zip file somewhere that has all the binaries, both dlls and the ikvmc.exe?  We're targeting .Net Framework 4.8 and have been using the ikvm 7 for years but need Java 8 support now.

Thanks

**Comment by NightOwl888** [comment]
We have released our first preview version of IKVM with .NET Core support. Please run it through its paces and report any issues you are seeing.

https://www.nuget.org/packages/IKVM/8.2.0-prerelease0379

**Comment by burakglc** [comment]
Hello.

Version 8.2.0-prerelease0379

cipher cipher = null;,

I define it as Cipher.getInstance("AES/ECB/PKCS5Padding"); I am getting error in method.
the error message is "The type initializer for 'javax.crypto.JceSecurity' threw an exception.' shaped.

**Comment by NightOwl888** [comment]
> Hello.
> 
> Version 8.2.0-prerelease0379
> 
> cipher cipher = null;,
> 
> I define it as Cipher.getInstance("AES/ECB/PKCS5Padding"); I am getting error in method. the error message is "The type initializer for 'javax.crypto.JceSecurity' threw an exception.' shaped.

Could you please open a new issue with that info so we can track it separately?

**Comment by burakglc** [comment]
A new issue has been opened.

**Comment by burakglc** [comment]
https://github.com/ikvm-revived/ikvm/issues/70

**Comment by gglyggly** [comment]
First, thanks very much for your work on the .net core version. I wonder where i can find the ikvmc in prerelease?

**Comment by NightOwl888** [comment]
All of the tools are available in `.zip` files on the [Releases page](https://github.com/ikvm-revived/ikvm/releases). The DLL references are available on [NuGet](https://www.nuget.org/packages/IKVM).

**Comment by gglyggly** [comment]
Thanks, the tools has ikvmc.exe, however, when I launched it, it has below error, anyone knows how to do that? 

`C:\Downloads\IKVM-8.2.0-prerelease.809-tools-netcoreapp3.1-win7-x64>ikvmc -out:lm.dll lm.jar
IKVM.NET Compiler (8.2.0-prerelease.809)
Copyright c 2022 Jeroen Frijters, Windward Studios, Jerome Haltom, Shad Storhaug


*** INTERNAL COMPILER ERROR ***

PLEASE FILE A BUG REPORT FOR IKVM.NET WHEN YOU SEE THIS MESSAGE

ikvmc, Version=8.2.0.0, Culture=neutral, PublicKeyToken=13235d27fcbfff58
C:\Downloads\IKVM-8.2.0-prerelease.809-tools-netcoreapp3.1-win7-x64\
3.1.25 64-bit

System.IO.FileNotFoundException: netstandard
   at IKVM.Reflection.Universe.Load(String refname, Module requestingModule, Boolean throwOnError) in D:\a\ikvm\ikvm\IKVM.Reflection\Universe.cs:line 828
   at IKVM.Internal.AssemblyResolver.Init(Universe universe, Boolean nostdlib, IList`1 references, IList`1 userLibPaths) in D:\a\ikvm\ikvm\ikvmc\IKVM\Internal\AssemblyResolver.cs:line 93
   at ikvmc.IkvmcCompiler.Compile(String[] args) in D:\a\ikvm\ikvm\ikvmc\IkvmcCompiler.cs:line 178
   at ikvmc.IkvmcCompiler.Main(String[] args) in D:\a\ikvm\ikvm\ikvmc\IkvmcCompiler.cs:line 112`

**Comment by gglyggly** [comment]
Thanks, the tools has ikvmc.exe. However, when I launched it, it has below error, anyone knows how to resolve? Is it SDK version related? I google but with no clue...

> C:\Downloads\IKVM-8.2.0-prerelease.809-tools-netcoreapp3.1-win7-x64>ikvmc -out:lm.dll lm.jar
> IKVM.NET Compiler (8.2.0-prerelease.809)
> Copyright c 2022 Jeroen Frijters, Windward Studios, Jerome Haltom, Shad Storhaug
> 
> 
> *** INTERNAL COMPILER ERROR ***
> 
> PLEASE FILE A BUG REPORT FOR IKVM.NET WHEN YOU SEE THIS MESSAGE
> 
> ikvmc, Version=8.2.0.0, Culture=neutral, PublicKeyToken=13235d27fcbfff58
> C:\Downloads\IKVM-8.2.0-prerelease.809-tools-netcoreapp3.1-win7-x64\
> 3.1.25 64-bit
> 
> **System.IO.FileNotFoundException: netstandard**
>    at IKVM.Reflection.Universe.Load(String refname, Module requestingModule, Boolean throwOnError) in D:\a\ikvm\ikvm\IKVM.Reflection\Universe.cs:line 828
>    at IKVM.Internal.AssemblyResolver.Init(Universe universe, Boolean nostdlib, IList`1 references, IList`1 userLibPaths) in D:\a\ikvm\ikvm\ikvmc\IKVM\Internal\AssemblyResolver.cs:line 93
>    at ikvmc.IkvmcCompiler.Compile(String[] args) in D:\a\ikvm\ikvm\ikvmc\IkvmcCompiler.cs:line 178
>    at ikvmc.IkvmcCompiler.Main(String[] args) in D:\a\ikvm\ikvm\ikvmc\IkvmcCompiler.cs:line 112

**Comment by NightOwl888** [comment]
> Thanks, the tools has ikvmc.exe. However, when I launched it, it has below error, anyone knows how to resolve? Is it SDK version related? I google but with no clue...
> 
> > C:\Downloads\IKVM-8.2.0-prerelease.809-tools-netcoreapp3.1-win7-x64>ikvmc -out:lm.dll lm.jar
> > IKVM.NET Compiler (8.2.0-prerelease.809)
> > Copyright c 2022 Jeroen Frijters, Windward Studios, Jerome Haltom, Shad Storhaug
> > *** INTERNAL COMPILER ERROR ***
> > PLEASE FILE A BUG REPORT FOR IKVM.NET WHEN YOU SEE THIS MESSAGE
> > ikvmc, Version=8.2.0.0, Culture=neutral, PublicKeyToken=13235d27fcbfff58
> > C:\Downloads\IKVM-8.2.0-prerelease.809-tools-netcoreapp3.1-win7-x64
> > 3.1.25 64-bit
> > **System.IO.FileNotFoundException: netstandard**
> > at IKVM.Reflection.Universe.Load(String refname, Module requestingModule, Boolean throwOnError) in D:\a\ikvm\ikvm\IKVM.Reflection\Universe.cs:line 828
> > at IKVM.Internal.AssemblyResolver.Init(Universe universe, Boolean nostdlib, IList`1 references, IList`1 userLibPaths) in D:\a\ikvm\ikvm\ikvmc\IKVM\Internal\AssemblyResolver.cs:line 93
> > at ikvmc.IkvmcCompiler.Compile(String[] args) in D:\a\ikvm\ikvm\ikvmc\IkvmcCompiler.cs:line 178
> > at ikvmc.IkvmcCompiler.Main(String[] args) in D:\a\ikvm\ikvm\ikvmc\IkvmcCompiler.cs:line 112

Please report this as a [new issue](https://github.com/ikvm-revived/ikvm/issues/new) so we can track it individually.

**Comment by gglyggly** [comment]
here it is [#74](https://github.com/ikvm-revived/ikvm/issues/74)

**Comment by viceice** [comment]
Which is the latest release on github? From releases page it's not clear

![image](https://user-images.githubusercontent.com/1798109/174317639-3a0994df-4cf9-44ef-896e-030a4bb0300a.png)

**Comment by viceice** [comment]
it seems 809 is latest, but on nuget.org 911 is latest, so there are also some github releases missing.

https://www.nuget.org/packages/IKVM/8.2.0-prerelease0911

**Comment by NightOwl888** [comment]
Looks like the last 3 GitHub [releases](https://github.com/ikvm-revived/ikvm/releases) are still in "draft" status. They all  have the same "what's changed" section, though (all of them are pointing to prerelease.809 for the beginning of the changelog).

@wasabii - I know I can change them to "real" releases through the control panel, but is there some compelling reason why they are still draft releases? Like maybe you were planning to change the release notes first?

**Comment by viceice** [comment]
@NightOwl888 I created #84 to fix it

**Comment by wasabii** [comment]
Hiya.

I thought it would be good to do it manually, after verifying everything was okay.

**Comment by viceice** [comment]
@wasabii But it seems the verify step got forgotten for latest releases 🙃 

So i would suggest to publish them right away. Missing releases are more worth than wrong release notes 🤷‍♂️
